### PR TITLE
feat: address projection

### PR DESCRIPTION
**Pull request by mkazlauskas**
# Context

Handle and utxo/chain history projections will require storing key hashes (payment and stake credentials) in order to be able to query by it.

# Proposed Solution

4f211bee5a92d390b53a3db3054a110f040168b4

# Important Changes Introduced

See commit log: all other commits are general fixes and refactors

### CODE REVIEW

**Review by greatertomi — COMMENTED** [review]
LGTM

**Review by greatertomi — APPROVED** [review]

File: packages/projection-typeorm/test/operators/storeAddresses.test.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        registration.pointer
+      )
+        .toAddress()
+        .toBech32() as Cardano.PaymentAddress;
+      await projectEvtWithAddress(address, Cardano.BlockNo(stakeKeyRegistrationEvent.block.header.blockNo + 1));
+      const projectedAddress = await addressesRepo.findOne({ where: { address } });
+      expect(typeof projectedAddress!.stakingCredentialHash).toBe('string');
+    });
+
+    it('projects invalid pointers (pointing to nothing) as undefined staking credential', async () => {
```
Comment (rhyslbw) [inline packages/projection-typeorm/test/operators/storeAddresses.test.ts:172]:
```suggestion
    it('projects null pointers as undefined staking credential', async () => {
```

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Looks really nice @mkazlauskas, great work! Minor issue highlighted inline

File: packages/projection-typeorm/test/operators/storeAddresses.test.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    buffer = new TypeormStabilityWindowBuffer({ allowNonSequentialBlockHeights: true, logger });
+    await buffer.initialize(queryRunner);
+  });
+
+  afterEach(async () => {
+    await queryRunner.release();
+    buffer.shutdown();
+  });
+
+  it('inserts addresses with their type, payment credential and staking credential', async () => {
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/projection-typeorm/test/operators/storeAddresses.test.ts]: 081a07c463ed0c23e214fcc95911e81271071e07

File: packages/projection-typeorm/src/entity/Address.entity.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  /**
+   * Applicable only for base/grouped, enterprise and pointer addresses
+   */
+  paymentCredentialHash?: Hash28ByteBase16 | null;
+  @Column('char', { length: 56, nullable: true })
+  @Index()
+  /**
+   * Applicable only for base/grouped and pointer addresses
+   */
+  stakingCredentialHash?: Hash28ByteBase16 | null;
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/projection-typeorm/src/entity/Address.entity.ts]: 081a07c463ed0c23e214fcc95911e81271071e07

File: packages/projection-typeorm/test/operators/storeAddresses.test.ts (hunk 1 above)
Comment (mkazlauskas) [reply packages/projection-typeorm/test/operators/storeAddresses.test.ts]: 27e2ff443fd191109bdcda6b9d1595866170e893 changed the suggested message to 'stake credential'

**Review by rhyslbw — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]
Awesome work @mkazlauskas 🚀

**Pull request merged**